# Analysis scripts

*mtDNA_copyNumber_phylogenies.Rmd* - Rmarkdown file for the analysis of mitochondrial variation in foetal blood from Spencer-Chapman et al. (**Figure 1** & **Figure 2** & **Supplementary Figure 2**)

*mtDNA_copyNumber_blood_emily.Rmd* - Rmarkdown file for the analysis of mitochondrial variation in adult blood from Mitchell et al. (**Figure 3** & **Supplementary Figure 3** & **Supplementary Figure 4**)

*calculate_median_cov_mtDNA.R* - Script to calculate coverage from foetal and adult blood data.

*create_trinucleotide_plots_blood.R* - Script for the visualization of the trinucleotide context for mtDNA variants called with ShearwaterML. (**Supplementary Figure 1**)

*tree_mtDNA_blood_emily.R* - Script to map mitochondrial mutations onto the phylogenetic trees from adult blood. (**Supplementary Figure 5**)

*estimate_betaBinom_parameters.R* - Script using bbmle package for the assessment of parameters for beta-binomial sampling of mtDNA copy number (**Supplementary Figure 6**)
